File: mazes-for-programmers/python/cell.py
class Cell:
    def __init__(self, row, column):
        self.row = row
        self.column = column
        self.links = {}
        self.north = None
        self.south = None
        self.east = None
        self.west = None

    def link(self, cell, bidi=True):
        self.links[cell] = True
        if bidi:
            cell.link(self, False)
        return self

    def unlink(self, cell, bidi=True):
        del self.links[cell]
        if bidi:
            cell.unlink(cell, False)
        return self

    def links(self):
        return self.links.keys()

    def is_linked(self, cell):
        return cell in self.links

    def wall_size(self, cell):
        if cell is None:
            return 2
        if cell not in self.links:
            return 1
        return 0

    def neighbors(self):
        list = []
        if self.north is not None:
            list.append(self.north)
        if self.south is not None:
            list.append(self.south)
        if self.west is not None:
            list.append(self.west)
        if self.east is not None:
            list.append(self.east)
        return list

File: mazes-for-programmers/kotlin/src/main/kotlin/io/github/ocirne/mazes/Cell.kt
package io.github.ocirne.mazes

open class Cell(open val row: Int, open val column: Int) {

    val links: MutableMap<Cell, Boolean> = mutableMapOf()

    var north: Cell? = null
    var south: Cell? = null
    var east: Cell? = null
    var west: Cell? = null

    fun link(cell: Cell, bidi: Boolean = true) {
        links[cell] = true
        if (bidi)
            cell.link(this, false)
    }

    fun unlink(cell: Cell, bidi: Boolean = true) {
        links.remove(cell)
        if (bidi)
            cell.unlink(this, false)
    }

    fun links(): MutableSet<Cell> {
        return links.keys
    }

    fun isLinked(cell: Cell?): Boolean {
        return links.containsKey(cell)
    }

    open fun neighbors(): List<Cell?> {
        return arrayListOf(north, south, east, west).filterNotNull()
    }

    fun distances(): Distances {
        val distances = Distances(this)
        var frontier = mutableListOf(this)

        while (frontier.isNotEmpty()) {
            val newFrontier: MutableList<Cell> = mutableListOf()
            for (cell in frontier) {
                for (linked in cell.links()) {
                    if (distances[linked] != null) {
                        continue
                    }
                    distances[linked] = distances[cell]!! + 1
                    newFrontier.add(linked)
                }
            }
            frontier = newFrontier
        }
        return distances
    }
}